Efficient Implementation of Self (#93)
sampsyo/cs6120: content/blog/2019-10-28-self/index.md

+++
title = "An Efficient Implementation of Self"

[extra]
bio = """
  [Rachit Nigam](https://rachitnigam.com) is a second year PhD student interested in
  programming languages & computer architecture. In his free time, he
  [subtweets](https://twitter.com/notypes/status/1170037148290080771) his advisor and [annoys tenured professors](https://twitter.com/natefoster/status/1074401015565291520).
"""
[[extra.authors]]
name = "Rachit Nigam"
link = "https://rachitnigam.com"
+++

[An Efficient Implementation of Self, a Dynamically-Typed Object-Oriented
Language Based on Prototypes][paper] presents techniques for runtime

> compilation, now more commonly referred to as Just In Time (JIT) compilation,

Reviewer: ```suggestion
compilation, now more commonly referred to as just-in-time (JIT) compilation,
```

It's usually not capitalized, but it usually does have hyphens.

for the [Self][] programming language.

> We have developed and implemented techniques that double the performance of
> the dynamically-typed object oriented languages.

## The Challenge

The term "dynamic language" is mostly commonly associated with modern scripting
languages like Python and JavaScript. Self, a much older language developed

> at the famed Xerox PARC labs takes the philosophy of dynamism to a logical

Reviewer: ```suggestion
at the famed Xerox PARC labs, takes the philosophy of dynamism to a logical
```

Missing a closer for the parenthetical comma.

extreme -- *everything* in self is a message to an object. This includes
Java-like method calls on object *as well as* control structures like loops

> and conditionals.

Reviewer: It might be worth mentioning that JavaScript is directly inspired by Self (and Scheme), which is why there will be some suspicious similarities here.

For example, the Self condition `IfTrue:IfFalse:` is a method invocation on
the boolean object `true` and `false`. This means that a simple conditional:

```
if (x) 1 else 2
```

> in Self parlance (written with JavaScript-esque syntax) looks like this:

Reviewer: The above is a little hard to understand for two reasons:

- It's not clear which of these is supposed to be JavaScript-like. I think it's the second one. But then it's not 100% sure what "language" the first snippet is written in—I think it's just a notional, abstract language, but maybe it's worth mentioning that it's not supposed to be Self or Self-esque JS.
- In JS, at least `if (x) 1 else 2` isn't terribly meaningful because `if` isn't an expression, unlike `x ? 1 : 2`. Maybe that's fine, though—this does make sense in, like, Scala or Rust.

```
x.If({ true: () => 1, false: () => 2 })
```

> The runtime behavior of this program is invoking the method `IfTrue` on the

Reviewer: ```suggestion
The run-time behavior of this program is invoking the method `IfTrue` on the
```
https://capra.cs.cornell.edu/styleguide/#runtime

object `x`, which can be a `true` or `false` but is not required to be those,
and execute the "thunk" (a function with no argument) corresponding to the true
or the false branch.  Note that `x` is not required to be `true` or `false`.
Any Self object can define the `If` method and specify its conditional

> behavior.

Reviewer: This paragraph switches a few times between several names for the same method: `If`, `IfTrue`, and `IfTrue:IfFalse:`. Maybe pick one?

Considering this dynamism, a Self compiler must:

1. **Respect the semantics of method calls**. Restricting or specializing conditionals and primitives don't follow the spirit of the language.
2. **Provide interactive speed**. Recompiling after a small change is not an option because Self is meant for rapid exploration in a programming environment.
3. **Preserve stack traces**. Self supports extreme reflection and introspection. A programming environment must be debuggable.
4. **Generate fast code**. Interpreting the whole language is probably too slow.

## The Solution

While the paper goes into the nitty-gritty of object layouts and method
invocations, the essence of the paper can be summarized as:

>  When type information for an object is available, generate specialized code
>  and let inlining and compiler optimizations work their magic.

The ideas described in the paper have influenced the design of modern JITs for
JavaScript which use a lot of [similar features][jit] to improve JavaScript execution

> times.

Reviewer: Not just JavaScript compilers—basically every JIT in use today descends from this set of ideas. (Even ones for more static-seeming languages like Java and C#.)

### Object Layout

Self programs use object prototypes to describe inheritance relationships.
Unlike classes, which have _constructors_ used to build _instances_, Self uses
prototypes, which act as _exemplars_ for other objects. Creating a new object
from a prototype is as simple as _cloning_ it and setting its parent pointer to
the prototype. Changing the common behavior of all clones is as simple as
changing the behavior of a prototype. Since method and field lookups traverse
the parent hierarchy, clones can also override methods and fields of their
prototypes.

A naive layout scheme for objects would copy all fields from a prototype and
end up wasting a lot of space describing potentially shared behaviors. The Self
compiler minimizes space usage of _clones_ derived from the same prototype by

> using _clone families_. A cloned object only stores its modifiable in its

Reviewer: Missing noun after "modifiable."

> object and points to the clone family for its prototype. If the instance ever

Reviewer: I'm not 100% sure what "in its object" means—maybe introduce something like an "allocation" to mean the actual memory for laying out fields, as opposed to "object" which is more abstract?

overrides one of its methods, the creates a new clone family to preserve
semantics and propagate behavior changes to all clones of a prototype.

<img src="without-maps.png" alt="drawing" width="300"/>

> <img src="with-maps.png" alt="drawing" width="300"/>

Reviewer: These figures are a little context free—maybe include a sentence or two describing what we're looking for in these pictures?

The paper mentions that contemporary Smalltalk-80 compilers restricted the
customization of primitive methods and control structures to allow for
specialized code generation. Instead of imposing these restrictions, the Self
compiler will generate specialized code for every _receiver_ object at runtime.
This means that at the first invocation of `<` on an integer, the compiler will
specialize `<` with check to see if the type matches to an `integer` and
generate a single compare instruction when this is case. When the type of
receiver has not been encountered before, Self will default to a method send
to the object.

### Message Inlining and Splitting

Inlining remains one of the most crucial optimization in modern compilers,
enabling other optimizations to become more effective. However, inlining
becomes impossible in presence of dynamic method calls if the type of the
receiver object is unknown. This is an even bigger problem for Self since
most of program is a sequence of dynamic calls.

The Self compiler performs two optimizations, message _inlining_ and
_splitting_ to enable efficient execution. Message _inlining_ works in a similar

> fashion to customized compilation -- If the type of a receiver is know, either

Reviewer: ```suggestion
fashion to customized compilation -- if the type of a receiver is known, either
```

at initial compilation through a dataflow analysis or at runtime, inline the
method body at the call location. When type information for an object is lost
due to control flow splits, _message splitting_ generates specialized code for
possible receiver types and guards them using type tests. If the type of

> receiver matches, fast code can be executed.

Reviewer: To drive this point home, maybe it would be useful to include some example code to illustrate what the program looks like after these optimizations are applied. For example, you could show several imaginary assembly- or C-like listings depicting these versions of a program `x < y`:

- The obvious version, where the code looks up the string `lessThan` in the method dictionary for the object `x` and jumps to the corresponding code with argument `y`.
- The "customized compilation" version, with a type guard and a direct `call` to the `integer.lessThan` function with `y` as the argument.
- The "inlined" version, which still has the type guard but just uses a `lt` machine instruction or whatever to do the comparison—no call necessary.

For bonus points, you could even show imaginary code for the "split" version.

### Programming Environment Support

> The Self compiler support _incremental recompilation_ and _source-level

Reviewer: ```suggestion
The Self compiler supports _incremental recompilation_ and _source-level
```

debugging_ by keeping track of the provenances of various method
specializations in a map.

Incremental Recompilation occurs when the compiler observes a change in the
programming environment and invalidates compiled methods associated with the
affected data. At compile time, the compiler creates a dependency list encoding
the information required for specialization. For example, if a method was
inlined from a clone's parent, updating the parent pointer or the method body
in the parent should invalidate the specialized code. Since the compiler
selectively invalidates code, methods that weren't modified can still execute

> with specialized code.

Reviewer: This is the weirdest part of Self, and the one that has not stuck with us today: the interactive, live-editable code environment. It's maybe worth pointing out that this is the most anachronistic part of the paper as a result. V8 doesn't have an interactive, persistent, first-class IDE, for example, so it doesn't have to worry about this as much.

to [forgo][v8-no-tail] tail call optimization due to due concerns about being
unable to reconstruct stack frames. The programming environment needs to be
able to step through specialized code as if it was going through the normal
method call chain. Self appends debugging information to each compiled method,
allowing the environment to reconstruct the state of the stack.

## A SELFish Evaluation

The Self compiler was written in 33,000 lines of C++ code and ran on the Sun-3.
The authors wrote 9,000 lines of Self code to implement the object hierarchy
and prototype graphical user interface.

The quantitative evaluation compares Self running times to that of the fastest
Smalltalk-80 compiler at the time by translating the Stanford Integer
Benchmarks. The authors both did a straightforward transliteration of Self programs
(marked `Self`) and also rewrote the benchmarks in idiomatic Self (marked `Self'`).

<img src="eval.png" alt="drawing" width="400"/>

On average, Self was 2-3x faster than the Smalltalk implementation and 4x slower
than the original C programs.

The authors also describes a new metric to measure the performance of an object

> oriented programming language - _Millions of Messages per Second (MiMS)_ which

Reviewer: Elsewhere, you have used ` -- ` to separate thoughts, rather than just a single hyphen. Consider sticking to that or just using commas here.

> corresponds to MIPS. To compute the MiMS of a specific virtual machine, divide

Reviewer: How does it "correspond" to MIPS? The paper says it's "analogous" to MIPS, which I think is more accurate.
